### DISCUSSION TITLE
Okta/SSO

### DISCUSSION THREAD

**Discussion by vladfr1** [danny-avila/LibreChat#1750] [Feature Requests & Suggestions]
Great tool, the team is having a lot of fun with it!!  We got everything working with Azure AD.   

Any thoughts on a more generic SSO implementation that could support any auth platform? i.e. Okta (our particular use case)?

**Comment by danny-avila** [comment]
Implementing Auth0 would be pretty involved but not impossible. The original intent with not using Auth0 was to not exclude users in regions where Auth0 is not supported.

**Comment by danny-avila** [comment]
Revisiting this, FWIW, Okta is working as an OpenID option, i can write a guide soon

**Reply by MariusWilsch** [reply]
Pleas do!

**Reply by puneet-babbar** [reply]
Any updates on the Okta guide?

**Reply by ahaj-98** [reply]
hey @danny-avila okta works with openid authentication for me, but it has no way for now to assign admin roles like keycloak

**Comment by vladfr1** [comment]
That would be awesome.  I am happy to to be a tester.

**Comment by danny-avila** [comment]
A very quick guide while it's still fresh:

I started with a fresh trial account, and added one user besides myself.

**You can setup the app with [Okta CLI](https://cli.okta.com/) but you can also do this through Okta's dashboard. I used the CLI.**

### For the CLI, you need an API key. You can create one via **Security > API > Tokens**

![image](https://github.com/user-attachments/assets/134c63ad-796a-4dd5-9d92-ee58487c3caf)

### While we're here, we also need to make sure our **Authorization Server has an Access Policy**. Unless you are well familiar with this, this is what I did with the default server:

1. **Goto Authorization Servers tab, click "default" hyperlink**
![image](https://github.com/user-attachments/assets/55cb0809-de6c-4010-8b6b-6b143e67a9d8)

2. **Goto Access Policies and create a lenient policy, or however you see fit.**
![image](https://github.com/user-attachments/assets/ef1008ab-7f5d-4da6-a63d-df9b09895d4f)

### Back to the CLI

Start terminal from project directory. Assuming you already have an account, run `okta login`. Then, run `okta apps create`.

- Select the default app name (LibreChat), or change it as you see fit. 
- Type of Application: Choose Web and press Enter.
- For Framework of Application: Select Other. 
- Then, the Redirect URI should be `http://localhost:3080/oauth/openid/callback` 
- use `http://localhost:3080/login` for the Logout Redirect URI.

It should save an `.okta.env` file once complete:

```bash
Okta application configuration has been written to: /home/danny/LibreChat/.okta.env
````

```.env
export OKTA_OAUTH2_ISSUER="https://trial-ID.okta.com/oauth2/default"
export OKTA_OAUTH2_CLIENT_ID="someID"
export OKTA_OAUTH2_CLIENT_SECRET="someSECRET"
```

Now, in our LibreChat .env file, we use these values like this:

```.env
# OKTA TEST
OPENID_CLIENT_ID=someID
OPENID_CLIENT_SECRET=someSECRET
OPENID_ISSUER=https://trial-ID.okta.com/oauth2/default
OPENID_SESSION_SECRET=secret # you can change this to whatever you'd like, should be some "secret" value
OPENID_SCOPE="openid profile email"
OPENID_CALLBACK_URL=/oauth/openid/callback
```

In our LibreChat .env file, Double check these values correspond to your deployment URL:

```.env
DOMAIN_CLIENT=http://localhost:3080
DOMAIN_SERVER=http://localhost:3080
```

By default the Okta CLI creates an app that assigns to all users. You may need to review all the bells & whistles Okta offers for this and adjust to your needs.

That's it! You should be able to login with Okta now.

**Reply by wizedkyle** [reply]
Just adding to this when setting the OPENID_ISSUER for Okta if you want to use the organisation authorisation server the URL will be: `https://<orgname>.okta.com`. Custom authorisation servers are a licensed feature which some customers may not have.

You can confirm the issuer URL using the following URL and looking at the "issuer" property: `https://<orgname>.okta.com/.well-known/openid-configuration`

**Reply by chengxiangzhu** [reply]
OPENID_AUTH_ENABLED=true
OPENID_CLIENT_ID=
OPENID_CLIENT_SECRET=
OPENID_ISSUER=
OPENID_SESSION_SECRET=

After configuring these, click the "Continue with OpenID" button and jump to the URL: https://example.com/oauth/openid. The error message is:
Error: Unknown authentication strategy "openid"
at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:193:39)

**Reply by danny-avila** [reply]
Updating Instructions (docker):
https://www.librechat.ai/docs/local/docker#update-librechat

**Reply by chengxiangzhu** [reply]
After the update, the error message still appears.
Run the command "git log"
![librechat](https://github.com/user-attachments/assets/5143f620-8633-43b0-adda-ea37e91fd54d)

**Reply by chengxiangzhu** [reply]
Run the command“git pull”
Already up to date.

**Reply by abhishekag03** [reply]
@chengxiangzhu I'm facing the same error - "Error: Unknown authentication strategy "openid"'. I have confirmed that I'm on the latest release v0.7.9. 
Were you able to fix this? If yes, how?

**Reply by ahaj-98** [reply]
hey @danny-avila  okta works with openid authentication for me, but it has no way for now to assign admin roles like keycloak

**Comment by vladfr1** [comment]
Hi DannyDo you ever do any setup consulting?Vlad Friedman410-960-8520On Sep 25, 2024, at 9:38 AM, ***@***.*** ***@***.*** { font-family: "Cambria Math"; }
@font-face { font-family: Calibri; }
@font-face { font-family: Aptos; }
p.MsoNormal, li.MsoNormal, div.MsoNormal { margin: 0in; font-size: 12pt; font-family: Aptos, sans-serif; }
a:link, span.MsoHyperlink { color: blue; text-decoration: underline; }
span.EmailStyle20 { font-family: Aptos, sans-serif; color: windowtext; }
.MsoChpDefault { font-size: 11pt; }
@page WordSection1 { size: 8.5in 11in; margin: 1in; }
div.WordSection1 { page: WordSection1; }That would be awesome.  I am happy to to be a tester.  From: Danny Avila ***@***.***> Sent: Wednesday, September 25, 2024 9:26 AMTo: danny-avila/LibreChat ***@***.***>Cc: vladfr1 ***@***.***>; Author ***@***.***>Subject: Re: [danny-avila/LibreChat] Okta/SSO (Discussion #1750) Revisiting this, FWIW, Okta is working as an OpenID option, i can write a guide soon—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

**Reply by aitp-john** [reply]
@vladfr1 do you still need help with set up?
